daemon: stop setting container resources to zero

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

Many of the fields in LinuxResources struct are pointers to scalars for some reason, presumably to differentiate between set-to-zero and unset when unmarshaling from JSON, despite zero being outside the acceptable range for the corresponding kernel tunables. When creating the OCI spec for a container, the daemon sets the container's OCI spec CPUShares and BlkioWeight parameters to zero when the corresponding Docker container configuration values are zero, signifying unset, despite the minimum acceptable value for CPUShares being two, and BlkioWeight ten. This has gone unnoticed as runC does not distingiush set-to-zero from unset as it also uses zero internally to represent unset for those fields. However, kata-containers v3.2.0-alpha.3 tries to apply the explicit-zero resource parameters to the container, exactly as instructed, and fails loudly. The OCI runtime-spec is silent on how the runtime should handle the case when those parameters are explicitly set to out-of-range values and kata's behaviour is not unreasonable, so the daemon must therefore be in the wrong.

Translate unset values in the Docker container's resources HostConfig to omit the corresponding fields in the container's OCI spec when starting and updating a container in order to maximize compatibility with runtimes.

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- Improved compatibility with OCI runtimes which are more strict than runC about resource constraint parameter values

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
daemon: modernize oci_linux_test.go

Switch to using t.TempDir() instead of rolling our own.

Clean up mounts leaked by the tests as otherwise the tests fail due to
the leaked mounts because unlike the old cleanup code, t.TempDir()
cleanup does not ignore errors from os.RemoveAll.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
daemon: stop setting container resources to zero

Many of the fields in LinuxResources struct are pointers to scalars for
some reason, presumably to differentiate between set-to-zero and unset
when unmarshaling from JSON, despite zero being outside the acceptable
range for the corresponding kernel tunables. When creating the OCI spec
for a container, the daemon sets the container's OCI spec CPUShares and
BlkioWeight parameters to zero when the corresponding Docker container
configuration values are zero, signifying unset, despite the minimum
acceptable value for CPUShares being two, and BlkioWeight ten. This has
gone unnoticed as runC does not distingiush set-to-zero from unset as it
also uses zero internally to represent unset for those fields. However,
kata-containers v3.2.0-alpha.3 tries to apply the explicit-zero resource
parameters to the container, exactly as instructed, and fails loudly.
The OCI runtime-spec is silent on how the runtime should handle the case
when those parameters are explicitly set to out-of-range values and
kata's behaviour is not unreasonable, so the daemon must therefore be in
the wrong.

Translate unset values in the Docker container's resources HostConfig to
omit the corresponding fields in the container's OCI spec when starting
and updating a container in order to maximize compatibility with
runtimes.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented on daemon/oci_linux_test.go:
```diff
@@ -48,6 +51,18 @@ func setupFakeDaemon(t *testing.T, c *container.Container) *Daemon {
 		c.NetworkSettings = &network.Settings{Networks: make(map[string]*network.EndpointSettings)}
 	}
 
+	// HORRIBLE HACK: clean up shm mounts leaked by some tests. Otherwise the
+	// offending tests would fail due to the mounts blocking the temporary
+	// directory from being cleaned up.
```
I checked: only the tests are impacted. The shm mounts are not leaked on container exit—though I'm not sure where in the code they get unmounted.

thaJeztah replied:
I think I was running into this one as well in https://github.com/moby/moby/pull/45474#discussion_r1197048332
and https://github.com/moby/moby/pull/43346/files#diff-7044cd74671fe1aa629c3374b3b65fb57e7d47c6a9a20a23b539ee1425a88f2eL19-L52


(Not sure why I didn't use `c.ShmPath` though 🤔

thaJeztah commented on daemon/oci_linux.go:
```diff
@@ -988,6 +986,10 @@ func WithResources(c *container.Container) coci.SpecOpts {
 			},
 			Pids: getPidsLimit(r),
 		}
+		if r.BlkioWeight != 0 {
+			w := r.BlkioWeight
+			specResources.BlockIO.Weight = &w
+		}
```
Definitely not a blocker, but would it make sense to have a `getBlkioWeight()` func (like we do for the other options above?

thaJeztah commented on daemon/oci_linux.go:
```diff
@@ -988,6 +986,10 @@ func WithResources(c *container.Container) coci.SpecOpts {
+			specResources.BlockIO.Weight = &w
+		}
 
 		if s.Linux.Resources != nil && len(s.Linux.Resources.Devices) > 0 {
 			specResources.Devices = s.Linux.Resources.Devices
```
Looking at this, I feel like `WithResources()` is doing too much, or we should split out a `WithBlkIO()` - it looks like this part is here, to prevent `WithResources()` from overwriting `Devices` (which probably is set separately), so trying to merge multiple options.

corhere replied:
No, ensuring that mutations to the spec by `WithDevices()` are not clobbered is _not_ `WithResources()` doing too much. It's just `WithResources()` being a good(-ish) implementation of a functional option, and what allows the functional options pattern to be used to factor distinct concerns into distinct option functions. Factoring out a `WithBlkIO()` function would also have to ensure that it does not clobber changes to the spec applied by `WithResources()` and other options.

That all being said, the way that `WithResources()` goes about not clobbering existing `LinuxResources` fields it does not otherwise touch is not particularly forward-compatible as it will zero out all other fields aside from `Devices`. I'll fix that.

thaJeztah commented on libcontainerd/types/types_linux.go:
```diff
@@ -27,7 +27,7 @@ func InterfaceToStats(read time.Time, v interface{}) *Stats {
 
 // Resources defines updatable container resource values. TODO: it must match containerd upcoming API
-type Resources specs.LinuxResources
+type Resources = specs.LinuxResources
```
Was this needed for this PR? I feel like this should be separate (and possibly when doing so, mark `Resources` deprecated in favour of `specs.LinuxResources`

corhere replied:
It avoids having to explicitly cast `types.Resources` values to `specs.LinuxResources`. And the type serves a purpose: it allows the libcontainerd `Task` interface to be have a single definition which is applicable for both Windows and Linux platforms. https://github.com/moby/moby/blob/1aae94074e63736a4b419f08cddc1baaaa717453/libcontainerd/types/types_windows.go#L26-L27
https://github.com/moby/moby/blob/1aae94074e63736a4b419f08cddc1baaaa717453/libcontainerd/types/types.go#L96

I am sure it would have been written as a type alias originally, if it existed at the time. (Proper type aliases were added in Go 1.9.) https://github.com/moby/moby/blob/1aae94074e63736a4b419f08cddc1baaaa717453/libcontainerd/remote/client_linux.go#L24-L26

I can split this out into its own commit if you feel strongly about it, but I don't think it's a significant enough change to warrant a separate PR. While technically a breaking change in the strictest sense, even Go's own apidiff tooling [does not consider merging types to be worth flagging.](https://pkg.go.dev/golang.org/x/exp/apidiff#section-readme:~:text=type%20potentially%20incompatible.-,Type%20Switches,-A%20package%20change)

thaJeztah reviewed (changes requested):
overall looks good; one comment on the type change for `Resources`

thaJeztah commented:
Actually; looks like there's some test-failures that may be related 😅

corhere committed:
daemon: ensure OCI options play nicely together

Audit the OCI spec options used for Linux containers to ensure they are
less order-dependent. Ensure they don't assume that any pointer fields
are non-nil and that they don't unintentionally clobber mutations to the
spec applied by other options.

Signed-off-by: Cory Snider <csnider@mirantis.com>

cpuguy83 reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM
